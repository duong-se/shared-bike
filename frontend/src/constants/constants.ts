export const tokenKey = 'accessToken'

export const commonHeaders = {
  'Accept': 'application/json',
  'Content-Type': 'application/json',
}
