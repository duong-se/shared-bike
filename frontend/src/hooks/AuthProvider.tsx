import { createContext, useCallback, useContext, useState } from "react"
import { useCookies } from 'react-cookie'

type AuthContextType = {
  login: (username: string, password: string, callback: VoidFunction) => void;
  logout: (callback: VoidFunction) => void;
  error?: string
}

const AuthContext = createContext<AuthContextType>(null!);

export const useAuth = () => {
  return useContext(AuthContext);
}

export const AuthProvider = ({ children }: { children: React.ReactNode }) => {
  const [, setCookie] = useCookies()
  const [error, setError] = useState<string| undefined>(undefined)
  const login = useCallback(async (username: string, password: string, callback: VoidFunction) => {
    const loginUrl = `${window.sharedBike.config.baseUrl}/users/login`
    fetch(loginUrl, {
      method: "POST",
      credentials: "include",
      headers: {
        "Accept": "application/json",
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        username,
        password,
      })
    }).then((resp) => {
      resp.status === 204 ? callback() : setError("Login failed")
    }).catch((error) => setError(error))
  }, [])

  const logout = useCallback(() => {
   setCookie('session', null)
  }, [setCookie])


  return <AuthContext.Provider value={{ login, logout, error }}>{children}</AuthContext.Provider>;
}
