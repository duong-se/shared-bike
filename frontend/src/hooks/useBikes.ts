import { useState } from "react"

export enum BikeStatus {
  RENTED = "rented",
  AVAILABLE = "available"
}

type Bike = {
  id: number
  lat: string
  long: string
  status: BikeStatus
  userId?: number
  nameOfRenter?: string
  usernameOfRenter?: string
}

export const useBikes = () => {
  const [error, setError] = useState<string| undefined>(undefined)
  const [isFetchLoading, setLoading] = useState<boolean>(true)
  const [isActionLoading, setActionLoading] = useState<boolean>(true)
  const [bikes, setBikes] = useState<Bike[]>([])
  const fetchAllBikes = async () => {
    setLoading(true)
    const getBikesUrl = `${window.sharedBike.config.baseUrl}/bikes`
    fetch(getBikesUrl, {
      method: "GET",
      credentials: "include",
      headers: {
        "Accept": "application/json",
        "Content-Type": "application/json"
      },
    }).then(async (resp) => {
      setLoading(false)
      if (resp.ok) {
        const bikes = await resp.json()
        return setBikes(bikes)
      }
    }).catch((error) => {
      setLoading(false)
      setError(error)
    })
  }

  const updateListBike = async (bikeId: number, resp: Bike) => {
    const filteredBikes = bikes.filter((item) => item.id !== bikeId)
    setBikes([...filteredBikes, resp])
  }

  const returnBike = async (bikeId: number) => {
    setActionLoading(true)
    const returnBikeUrl = `${window.sharedBike.config.baseUrl}/bikes/${bikeId}/return`
    fetch(returnBikeUrl, {
      method: "PATCH",
      credentials: "include",
      headers: {
        "Accept": "application/json",
        "Content-Type": "application/json"
      },
    }).then(async (resp) => {
      setActionLoading(false)
      if (resp.ok) {
        const bikeUpdated = await resp.json()
        updateListBike(bikeId, bikeUpdated)
      }
    }).catch((error) => {
      setActionLoading(false)
      setError(error)
    })
  }

  const rentBike = async (bikeId: number) => {
    setActionLoading(true)
    const returnBikeUrl = `${window.sharedBike.config.baseUrl}/bikes/${bikeId}/rent`
    fetch(returnBikeUrl, {
      method: "PATCH",
      credentials: "include",
      headers: {
        "Accept": "application/json",
        "Content-Type": "application/json"
      },
    }).then(async (resp) => {
      setActionLoading(false)
      if (resp.ok) {
        const bikeUpdated = await resp.json()
        updateListBike(bikeId, bikeUpdated)
      }
    }).catch((error) => {
      setActionLoading(false)
      setError(error)
    })
  }

  return {
    fetchAllBikes,
    returnBike,
    rentBike,
    error,
    bikes,
    isFetchLoading,
    isActionLoading
  }
}
