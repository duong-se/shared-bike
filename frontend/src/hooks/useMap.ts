import { useCallback, useState } from 'react'

export type UseMapReturn = {
  mapRef: (node: HTMLDivElement | null) => void
  map?: google.maps.Map
}

export const useMap = (position: google.maps.LatLng): UseMapReturn => {
  const [map, setMap] = useState<google.maps.Map>()
  const mapRef = useCallback((node: HTMLDivElement | null) => {
    if (node !== null) {
      const map = new window.google.maps.Map(node, {
        center: position,
        zoom: 8,
        fullscreenControl: false,
      })
      setMap(map)
    }
  }, [])
  return {
    map,
    mapRef,
  }
}
