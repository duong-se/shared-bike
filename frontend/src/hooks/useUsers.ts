import { useMutation, UseMutationOptions } from 'react-query'
import { login, register } from '../apis/users'
import { LoginResponse, LoginVariables } from '../typings/types'
import { RegisterResponse, RegisterVariables } from '../typings/types'

export const useLogin = (options?: Omit<UseMutationOptions<LoginResponse, unknown, LoginVariables, unknown>, 'mutationFn'>) => {
  return useMutation(login, options)
}

export const useRegister = (options?: Omit<UseMutationOptions<RegisterResponse, unknown, RegisterVariables, unknown>, 'mutationFn'>) => {
  return useMutation(register, options)
}
