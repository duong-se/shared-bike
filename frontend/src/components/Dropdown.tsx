import { useCallback } from 'react'
import { useNavigate } from 'react-router-dom'
import { tokenKey } from '../constants/constants'
import { useAuth } from '../hooks/AuthProvider'

export const Dropdown: React.FC = () => {
  const { user } = useAuth()
  const navigation = useNavigate()
  const handleLogout = useCallback(() => {
    localStorage.removeItem(tokenKey)
    navigation('/', { replace: true })
  }, [navigation])

  if (!user) {
    return null
  }
  return (
    <div className="absolute z-10 right-1">
      <div className="dropdown dropdown-end">
        <label data-testid="avatar" tabIndex={0} className="btn btn-ghost btn-circle avatar placeholder">
          <div className="bg-neutral-focus text-neutral-content w-10 rounded-full text-xs">
            <span className="text-xs">{user.name[0]}</span>
          </div>
        </label>
        <ul tabIndex={0} className="menu menu-compact dropdown-content mt-3 p-2 shadow bg-base-100 rounded-box w-52">
          <li><button onClick={handleLogout}>Logout</button></li>
        </ul>
      </div>
    </div>
  )
}
