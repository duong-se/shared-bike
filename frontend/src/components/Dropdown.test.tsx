import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { tokenKey } from '../constants/constants'
import { AuthProvider } from '../hooks/AuthProvider'
import { Dropdown } from './Dropdown'

describe('Dropdown', () => {
  it('should render correctly and clear token when logout', () => {
    localStorage.setItem(tokenKey, 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MSwidXNlcm5hbWUiOiJ0ZXN0MSIsIm5hbWUiOiJUZXN0IFVzZXIiLCJwZXJtaXNzaW9ucyI6bnVsbCwiZXhwIjoxNjU3MzUyODM0fQ.jU5yp2y-3H-dxXP1hdDW-FYEYv5F9GhAVDCbafphUzI')
    const token = localStorage.getItem(tokenKey)
    expect(token).not.toBeNull()
    render(<AuthProvider><Dropdown /></AuthProvider>, { wrapper: BrowserRouter })
    const avatarButton = screen.getByTestId('avatar')
    expect(avatarButton).toBeInTheDocument()
    fireEvent.click(avatarButton)
    const dropdown = screen.getByRole('button')
    fireEvent.click(dropdown)
    waitFor(() => {
      const clearedToken = localStorage.getItem(tokenKey)
      expect(clearedToken).toBeNull()
    })
  })

  it('should render correctly null', () => {
    localStorage.clear()
    const token = localStorage.getItem(tokenKey)
    expect(token).toBeNull()
    const { container } = render(<AuthProvider><Dropdown /></AuthProvider>, { wrapper: BrowserRouter })
    expect(container).toMatchSnapshot()
  })
})
