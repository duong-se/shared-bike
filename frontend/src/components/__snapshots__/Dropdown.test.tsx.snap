// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render correctly null 1`] = `<div />`;
